marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/getDocumentForTesting.sjs

```diff
@@ -19,13 +19,83 @@ xdmp.securityAssert('http://marklogic.com/data-hub/privileges/read-mapping', 'ex
 
 const core = require('/data-hub/5/artifacts/core.sjs')
 const ds = require("/data-hub/5/data-services/ds-utils.sjs");
+const xmlToJson = require('./xmlToJsonForMapping.sjs');
 
 var stepName, uri;
 
+function _isSourceJson(format) {
+  return format.toUpperCase() === 'JSON'
+}
+
+function _isValidQName(name) {
+  try {
+    fn.QName('', name);
+    return true;
+  } catch (e) {
+    return false;
+  }
+}
+
+function _isObject(value) {
+  // Added key criteria as typeof(value) returned 'object' for some scalar values.
+  return (value && typeof(value) === 'object' && Object.keys(value).length > 0) === true;
+}
+
+function _isAtomic(value) {
+  return !_isObject(value);
+}
+
+function _isArray(value) {
+  // False negatives from Array.isArray(value)
+  return value && value.hasOwnProperty('0');
+}
+
+function _getXPath(leadingPath, nextPart, value, format, isArray) {
+  // Account for invalid QNames, which is possible when the source format is JSON.
+  let funcStart = '';
+  let funcEnd = '';
+  if (_isSourceJson(format) && !_isValidQName(nextPart)) {
+    // Array of atomic values
+    if (isArray && value.length > 0 && _isAtomic(value[0])) {
+      funcStart = "array-node('";
+      funcEnd = "')/node()";
+    } else {
+      // Either not an array, an empty array, or an array of object values.
+      funcStart = "node('";
+      funcEnd = "')";
+    }
+  }
+  return `${leadingPath}/${funcStart}${nextPart}${funcEnd}`;
+}
+
+// Recursive function used to populate the sourceProperties portion/array of the return.
+function _flatten(sourceData, sourceFormat, flatArray, flatArrayKey = '', level = 0) {
```

Comment:
Assuming all of the functions added here are needed for this "flatten" function, let's move them into a separate library (in the same directory). That allows us to export the one public method that's needed here, and to avoid using "_", which is not a convention that's used anywhere else in DHF. 

"flatten" doesn't seem like the right function name either, as that's understood in JS to mean what "Array.flat()" means. And the comment indicates that it does something more interesting too - to populate the sourceProperties object. 

The user goal appears to be - given a source document and a source format, I want back an array with the source properties that I find in the source document. Right? So I think that public function in the library should be:

    buildSourceProperties(sourceDocument, sourceRecord)

Which would return an array. And then that library can have a recursive function that has a signature that includes the array, the array key, and the level.

Reply:
Implemented as requested.  The relocated code is within `testingSourceProperties.sjs` (open to a different name).